[Clang] Add GCC's __builtin_stack_address() to Clang.

aalhwc opened the pull request:
This new builtin returns the value of the stack pointer register, mirroring GCC's [__builtin_stack_address()](https://gcc.gnu.org/onlinedocs/gcc/Return-Address.html). This implementation initially supports the x86, x86_64, risc32, risc64, arm32, arm64 architectures. Support for other architectures can be added in future patches.

Referring: #82632

aalhwc committed:
[Clang] Add GCC's __builtin_stack_address() to Clang (#82632).

This new builtin returns the value of the stack pointer register,
mirroring GCC's __builtin_stack_address(). This implementation initially
supports only the x86 and x86_64 architectures. Support for other
architectures can be added in future patches.

tbaederr commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -4782,6 +4782,30 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
+      SPRegName = "rsp";
+      break;
+    default:
+      llvm_unreachable("Intrinsic __builtin_stack_address is not supported for "
+                       "the target architecture");
+    }
```
Where is it ensured that this is actually unreachable?

aalhwc replied:
Maybe  `llvm_unreachable` is inappropriate in this case.

The purpose of that `llvm::unreachable` statement is to indicate that  `__builtin_stack_address` is either: (1) supported by the target architecture but not yet implemented in LLVM (I plan to add support for more architectures in future patches) or (2) not supported by the target architecture (e.g. there's no stack pointer register). The `llvm_unreachable` will fire when either (1) or (2) is true.

Would it be more appropriate to `llvm::report_fatal_error` or simply an `assert` in this case?

philnik777 replied:
You should probably make this a target-specific builtin. The compiler shouldn't crash just because someone used an attribute on a platform where it's not supported.

aalhwc replied:
> The compiler shouldn't crash just because someone used an attribute on a platform where it's not supported.

I agree. It's probably more appropriate to report an error instead of crashing?

> You should probably make this a target-specific builtin

I see.  I made this a general builtin because I noticed that `__builtin_frame_address` and `__builtin_return_address` are defined as general builtin's so I thought `__builtin_stack_address` should follow the same approach?

Making this a target-specific builtin is a good idea, but what do you think about keeping `__builtin_stack_address` as a general builtin and generating a [`Cannot compile this __builtin_stack_address yet` compile error](https://github.com/llvm/llvm-project/blob/d5c8af492f2d8620b04330024d46a5f48db546fe/clang/lib/CodeGen/CGBuiltin.cpp#L4800)) when the target arch is unsupported? The compile error will look like the following:
```cpp
llvm-project/clang/test/CodeGen/builtin-stackaddress.c:17:10: error: cannot compile this __builtin_stack_address yet
   17 |   return __builtin_stack_address();
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
 Insights are appreciated!

aalhwc replied:
I updated the implementation to produce a compile error when the target architecture is unsupported.

I reused the existing function [`EmitSpecialRegisterBuiltin`](https://github.com/llvm/llvm-project/blob/2fae5bdea7c2016d4086aa7ecf3c5d0592ce95c8/clang/lib/CodeGen/CGBuiltin.cpp#L8929) to generate the `@llvm.read_register` intrinsic. That required moving it earlier in the source file.

I can make this a target-specific builtin if this implementation is not appropriate.

aalhwc committed:
[Clang] Add RISCV support for `__builtin_stack_address`.

Read the 'sp' register, replicating GCC's implementation of
`__builtin_stack_address` on RISCV: https://godbolt.org/z/fGraxYoxd

aalhwc committed:
[Clang] Add ARM support for `__builtin_stack_address`.

Read the 'sp' register, replicating GCC's implementation of
`__builtin_stack_address` on ARM: https://godbolt.org/z/4EY531j89

aalhwc commented:
@philnik777 does this new version of the patch address your comment?

aalhwc commented:
Ping

aalhwc commented:
> Where is it ensured that this is actually unreachable?

@tbaederr I now report a compile time error instead of crashing. Does that address your comment appropriately? Thanks!

AaronBallman commented on clang/include/clang/Basic/Builtins.td:
```diff
@@ -899,6 +899,12 @@ def FrameAddress : Builtin {
 
+def StackAddress : Builtin {
+  let Spellings = ["__builtin_stack_address"];
+  let Attributes = [NoThrow];
```
Should this be marked `Const` (or maybe `Pure`)?

aalhwc replied:
I may be wrong here but I am not sure if this is an instance of a `Const` or even `Pure` function. Two consecutive calls to `__builtin_stack_address` can return different values which makes it im`Pure`. Since `Const` functions are a subset of `Pure` functions, it also can't be `Const`. This is based on [GCC](https://gcc.gnu.org/onlinedocs/gcc-4.0.0/gcc/Function-Attributes.html)'s documentation regarding function attributes.

AaronBallman replied:
I wasn't 100% sure; I thought it meant two consecutive calls from the same context, but I can never remember that detail. I'm fine leaving it off though -- we can easily add it later if we find there's a benefit. Thanks!

AaronBallman commented on clang/test/CodeGen/builtin-stackaddress.c:
```diff
@@ -0,0 +1,25 @@
+  // RISCV_ARM_64: [[INT_SP:%.*]] = call i64 @llvm.read_register.i64(metadata [[SPREG:![0-9]+]])
+  // RISCV_ARM_64: inttoptr i64 [[INT_SP]]
+  // RISCV_ARM_64: [[SPREG]] = !{!"sp"}
+  return __builtin_stack_address();
```
We definitely need some tests in `clang/test/Sema` for things like diagnosing passing arguments to the function and verifying the function's return type.

But also, we may want codegen tests for more interesting scenarios (member functions in C++, lambdas in C++, functions marked with the `leaf` attribute, etc).

aalhwc replied:
Noted! Thanks.

AaronBallman commented on clang/lib/CodeGen/CGBuiltin.cpp:
```diff
@@ -4782,6 +4857,34 @@ RValue CodeGenFunction::EmitBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
+          *this, E, SPRegIntTy, VoidPtrTy, NormalRead, "sp"));
+    }
+    default:
+      ErrorUnsupported(E, "__builtin_stack_address");
```
We're missing test coverage for unsupported architectures; this seems like something that should be rejected before CodeGen too (in SemaChecking.cpp)

aalhwc replied:
Noted! Thanks!

AaronBallman reviewed (commented):
Thank you for this!

Please also be sure to add a release note to `clang/docs/ReleaseNotes.rst` as well as document the extension in `clang/docs/LanguageExtensions.rst` so users know about the functionality.

Should we be supporting `STACK_ADDRESS_OFFSET` similar to what GCC does?

nikic reviewed (commented):
Other builtins like __builtin_frame_address and __builtin_return_address map onto an llvm intrinsic like llvm.frameaddress or llvm.returnaddress. What's the reason for not following this design for __builtin_stack_address?

rjmccall commented:
I agree with Nikita. I have no problem with supporting this builtin, but it feels like something that should be expressed portably as an IR intrinsic and then get lowered in the backend rather than needing frontend work for every target.

aalhwc commented:
Thanks for all the comments!

> Should we be supporting STACK_ADDRESS_OFFSET similar to what GCC does?

Apologies for missing that. I'll have it ready before the next patch update.

> Other builtins [...] map onto an llvm intrinsic [...] What's the reason for not following this design for __builtin_stack_address?

> it feels like something that should be expressed portably as an IR intrinsic and then get lowered in the backend [...]

I wasn't aware of the direct mapping between the front end intrinsics and the llvm intrinsics. Implementing the llvm intrinsic `llvm.stackaddress` makes perfect sense. I'll try to have it ready before the next patch update. Thanks!
